File: book/petrol-pump/java/chapter4/section11/Preset.java
package chapter4.section11;

import pump.*;
import chapter4.section7.Fill;
import nz.sodium.*;
import java.util.Optional;

public class Preset {
    public final Cell<Delivery> delivery;
    public final Cell<Boolean> keypadActive;

    public enum Speed { FAST, SLOW, STOPPED };

    public Preset(Cell<Integer> presetDollars,
                  Fill fi,
                  Cell<Optional<Fuel>> fuelFlowing,
                  Cell<Boolean> fillActive) {
        Cell<Speed> speed = presetDollars.lift(
        	    fi.price, fi.dollarsDelivered, fi.litersDelivered,
            (presetDollars_, price, dollarsDelivered, litersDelivered) -> {
                if (presetDollars_ == 0)
                    return Speed.FAST;
                else {
                    if (dollarsDelivered >= (double)presetDollars_)
                        return Speed.STOPPED;
                    double slowLiters =
                            (double)presetDollars_/price - 0.10;
                    if (litersDelivered >= slowLiters)
                        return Speed.SLOW;
                    else
                        return Speed.FAST;
                }
            });
        delivery = fuelFlowing.lift(speed,
            (of, speed_) ->
                speed_ == Speed.FAST ? (
                    of.equals(Optional.of(Fuel.ONE))   ? Delivery.FAST1 :
                    of.equals(Optional.of(Fuel.TWO))   ? Delivery.FAST2 :
                    of.equals(Optional.of(Fuel.THREE)) ? Delivery.FAST3 : 
                                                         Delivery.OFF
                ) :
                speed_ == Speed.SLOW ? (
                    of.equals(Optional.of(Fuel.ONE))   ? Delivery.SLOW1 :
                    of.equals(Optional.of(Fuel.TWO))   ? Delivery.SLOW2 :
                    of.equals(Optional.of(Fuel.THREE)) ? Delivery.SLOW3 : 
                                                         Delivery.OFF
                ) :
                Delivery.OFF);
        keypadActive = fuelFlowing.lift(speed,
            (of, speed_) ->
                !of.isPresent() || speed_ == Speed.FAST);
    }
}

File: book/petrol-pump/c#/PetrolPump/Chapter4/Section11/Preset.cs
using PetrolPump.Chapter4.Section7;
using Sodium.Frp;
using Sodium.Functional;

namespace PetrolPump.Chapter4.Section11
{
    internal class Preset
    {
        public Cell<Delivery> Delivery { get; }
        public Cell<bool> KeypadActive { get; }

        private enum Speed
        {
            Fast,
            Slow,
            Stopped
        };

        public Preset(Cell<int> presetDollars, Fill fi, Cell<Maybe<Fuel>> fuelFlowing)
        {
            Cell<Speed> speed = presetDollars.Lift(
                fi.Price, fi.DollarsDelivered, fi.LitersDelivered,
                (presetDollarsLocal, price, dollarsDelivered, litersDelivered) =>
                {
                    if (presetDollarsLocal == 0)
                    {
                        return Speed.Fast;
                    }

                    if (dollarsDelivered >= presetDollarsLocal)
                    {
                        return Speed.Stopped;
                    }

                    double slowLiters = presetDollarsLocal / price - 0.10;
                    return litersDelivered >= slowLiters ? Speed.Slow : Speed.Fast;
                });
            this.Delivery = fuelFlowing.Lift(speed,
                (m, speedLocal) =>
                    speedLocal == Speed.Fast ? (
                        m.Equals(Maybe.Some(Fuel.One)) ? PetrolPump.Delivery.Fast1 :
                            m.Equals(Maybe.Some(Fuel.Two)) ? PetrolPump.Delivery.Fast2 :
                                m.Equals(Maybe.Some(Fuel.Three)) ? PetrolPump.Delivery.Fast3 : PetrolPump.Delivery.Off
                        ) :
                        speedLocal == Speed.Slow ? (
                            m.Equals(Maybe.Some(Fuel.One)) ? PetrolPump.Delivery.Slow1 :
                                m.Equals(Maybe.Some(Fuel.Two)) ? PetrolPump.Delivery.Slow2 :
                                    m.Equals(Maybe.Some(Fuel.Three)) ? PetrolPump.Delivery.Slow3 : PetrolPump.Delivery.Off
                            ) : PetrolPump.Delivery.Off);
            this.KeypadActive = fuelFlowing.Lift(speed, (m, speedLocal) => m.Match(_ => speedLocal == Speed.Fast, () => true));
        }
    }
}